<h1><center>Adults Dataset</center></h1>

# Machine Learning 1 - Lab 2: Classification

### Authors:
Richard Kim <br> Connor Dobbs<br> Joaquin Dominguez

## Data Preparation, Part 1

Define and prepare your class variables. Use proper variable representations (int, float, one-hot, etc.). Use pre-processing methods (as needed) for dimensionality reduction, scaling, etc. Remove variables that are not needed/useful for the analysis.

### Import packages and Dataset

In [56]:
# Import relevant packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Hide deprecation warnings
import warnings
warnings.filterwarnings('ignore', category=FutureWarning)
warnings.filterwarnings('ignore', category=DeprecationWarning)

In [57]:
### Import dataset into 'df'
df = pd.read_csv('https://raw.githubusercontent.com/j-dominguez9/ML1_Proj1/main/Data/census_clean.csv')

In [58]:
## check types, get overview
df.info()

#double check nulls
df.isna().sum()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45175 entries, 0 to 45174
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Unnamed: 0       45175 non-null  int64 
 1   age              45175 non-null  int64 
 2   workclass        45175 non-null  object
 3   fnlwgt           45175 non-null  int64 
 4   education        45175 non-null  object
 5   educational-num  45175 non-null  int64 
 6   marital-status   45175 non-null  object
 7   occupation       45175 non-null  object
 8   relationship     45175 non-null  object
 9   race             45175 non-null  object
 10  gender           45175 non-null  object
 11  capital-gain     45175 non-null  int64 
 12  capital-loss     45175 non-null  int64 
 13  hours-per-week   45175 non-null  int64 
 14  native-country   45175 non-null  object
 15  income           45175 non-null  object
dtypes: int64(7), object(9)
memory usage: 5.5+ MB


Unnamed: 0  age  workclass  fnlwgt     education  educational-num  \
0           0   25    Private  226802          11th                7   
1           1   38    Private   89814       HS-grad                9   
2           2   28  Local-gov  336951    Assoc-acdm               12   
3           3   44    Private  160323  Some-college               10   
4           5   34    Private  198693          10th                6   

       marital-status         occupation   relationship   race gender  \
0       Never-married  Machine-op-inspct      Own-child  Black   Male   
1  Married-civ-spouse    Farming-fishing        Husband  White   Male   
2  Married-civ-spouse    Protective-serv        Husband  White   Male   
3  Married-civ-spouse  Machine-op-inspct        Husband  Black   Male   
4       Never-married      Other-service  Not-in-family  White   Male   

   capital-gain  capital-loss  hours-per-week native-country income  
0             0             0              40  United-States  <=50K  
1             0             0              50  United-States  <=50K  
2             0             0              40  United-States   >50K  
3          7688             0              40  United-States   >50K  
4             0             0              30  United-States  <=50K

### Check Distribution of Target Variable

In [59]:
# check balancing
df['income'].value_counts()

<=50K    33973
>50K     11202
Name: income, dtype: int64

In [60]:
# get percentages for 'income' distribution
less = 33973
greater = 11202
tot = less + greater
print(f'<=50k: {(less/tot)*100}%')
print(f'>50k: {(greater/tot)*100}%')


<=50k: 75.20309905921417%
>50k: 24.796900940785832%


It is clear from the distribution of our target variable (as outlined above) that applying a sampling method to balance the dataset would benefit the models that follow from the dataset.

### Prepare Dataset for Classification

In this step, we transformed all categorical variables (via OneHotEncoding method) to prepare the data for logistic regression. Note that in further modeling, these categorical variables will be refined further. 

In [61]:
# create dummy var for income

dummy = pd.get_dummies(df['income'])

df_1 = pd.concat((df, dummy), axis=1)

df_1 = df_1.drop(['income'], axis=1)

df_1 = df_1.drop(['<=50K'], axis=1)

df_1 = df_1.rename(columns={'>50K': 'income'})

In [62]:
# OneHotEncoding workclass variable
dummy = pd.get_dummies(df['workclass'])

df_1 = pd.concat((df_1, dummy), axis=1)

df_1 = df_1.drop(['workclass'], axis=1)

df_1.head()

Unnamed: 0  age  fnlwgt     education  educational-num      marital-status  \
0           0   25  226802          11th                7       Never-married   
1           1   38   89814       HS-grad                9  Married-civ-spouse   
2           2   28  336951    Assoc-acdm               12  Married-civ-spouse   
3           3   44  160323  Some-college               10  Married-civ-spouse   
4           5   34  198693          10th                6       Never-married   

          occupation   relationship   race gender  ...  hours-per-week  \
0  Machine-op-inspct      Own-child  Black   Male  ...              40   
1    Farming-fishing        Husband  White   Male  ...              50   
2    Protective-serv        Husband  White   Male  ...              40   
3  Machine-op-inspct        Husband  Black   Male  ...              40   
4      Other-service  Not-in-family  White   Male  ...              30   

   native-country  income Federal-gov  Local-gov  Private  Self-emp-inc  \
0   United-States       0           0          0        1             0   
1   United-States       0           0          0        1             0   
2   United-States       1           0          1        0             0   
3   United-States       1           0          0        1             0   
4   United-States       0           0          0        1             0   

   Self-emp-not-inc  State-gov  Without-pay  
0                 0          0            0  
1                 0          0            0  
2                 0          0            0  
3                 0          0            0  
4                 0          0            0  

[5 rows x 22 columns]

In [63]:
# OneHotEncoding gender variable
dummy = pd.get_dummies(df['gender'])

df_1 = pd.concat((df_1, dummy), axis=1)

df_1 = df_1.drop(['gender'], axis=1)

df_1.head()

Unnamed: 0  age  fnlwgt     education  educational-num      marital-status  \
0           0   25  226802          11th                7       Never-married   
1           1   38   89814       HS-grad                9  Married-civ-spouse   
2           2   28  336951    Assoc-acdm               12  Married-civ-spouse   
3           3   44  160323  Some-college               10  Married-civ-spouse   
4           5   34  198693          10th                6       Never-married   

          occupation   relationship   race  capital-gain  ...  income  \
0  Machine-op-inspct      Own-child  Black             0  ...       0   
1    Farming-fishing        Husband  White             0  ...       0   
2    Protective-serv        Husband  White             0  ...       1   
3  Machine-op-inspct        Husband  Black          7688  ...       1   
4      Other-service  Not-in-family  White             0  ...       0   

   Federal-gov Local-gov  Private  Self-emp-inc  Self-emp-not-inc  State-gov  \
0            0         0        1             0                 0          0   
1            0         0        1             0                 0          0   
2            0         1        0             0                 0          0   
3            0         0        1             0                 0          0   
4            0         0        1             0                 0          0   

   Without-pay  Female  Male  
0            0       0     1  
1            0       0     1  
2            0       0     1  
3            0       0     1  
4            0       0     1  

[5 rows x 23 columns]

In [64]:
# OneHotEncoding education var
dummy = pd.get_dummies(df['education'])

df_1 = pd.concat((df_1, dummy), axis=1)

df_1 = df_1.drop(['education'], axis=1)

df_1.head()

Unnamed: 0  age  fnlwgt  educational-num      marital-status  \
0           0   25  226802                7       Never-married   
1           1   38   89814                9  Married-civ-spouse   
2           2   28  336951               12  Married-civ-spouse   
3           3   44  160323               10  Married-civ-spouse   
4           5   34  198693                6       Never-married   

          occupation   relationship   race  capital-gain  capital-loss  ...  \
0  Machine-op-inspct      Own-child  Black             0             0  ...   
1    Farming-fishing        Husband  White             0             0  ...   
2    Protective-serv        Husband  White             0             0  ...   
3  Machine-op-inspct        Husband  Black          7688             0  ...   
4      Other-service  Not-in-family  White             0             0  ...   

   9th Assoc-acdm  Assoc-voc  Bachelors  Doctorate  HS-grad  Masters  \
0    0          0          0          0          0        0        0   
1    0          0          0          0          0        1        0   
2    0          1          0          0          0        0        0   
3    0          0          0          0          0        0        0   
4    0          0          0          0          0        0        0   

   Preschool  Prof-school  Some-college  
0          0            0             0  
1          0            0             0  
2          0            0             0  
3          0            0             1  
4          0            0             0  

[5 rows x 38 columns]

In [65]:
#OneHotEncoding marital-status var
dummy = pd.get_dummies(df['marital-status'])

df_1 = pd.concat((df_1, dummy), axis=1)

df_1 = df_1.drop(['marital-status'], axis=1)

df_1.head()

Unnamed: 0  age  fnlwgt  educational-num         occupation   relationship  \
0           0   25  226802                7  Machine-op-inspct      Own-child   
1           1   38   89814                9    Farming-fishing        Husband   
2           2   28  336951               12    Protective-serv        Husband   
3           3   44  160323               10  Machine-op-inspct        Husband   
4           5   34  198693                6      Other-service  Not-in-family   

    race  capital-gain  capital-loss  hours-per-week  ... Preschool  \
0  Black             0             0              40  ...         0   
1  White             0             0              50  ...         0   
2  White             0             0              40  ...         0   
3  Black          7688             0              40  ...         0   
4  White             0             0              30  ...         0   

   Prof-school  Some-college  Divorced  Married-AF-spouse  Married-civ-spouse  \
0            0             0         0                  0                   0   
1            0             0         0                  0                   1   
2            0             0         0                  0                   1   
3            0             1         0                  0                   1   
4            0             0         0                  0                   0   

   Married-spouse-absent  Never-married  Separated  Widowed  
0                      0              1          0        0  
1                      0              0          0        0  
2                      0              0          0        0  
3                      0              0          0        0  
4                      0              1          0        0  

[5 rows x 44 columns]

In [66]:
#OneHotEncoding occupation var
dummy = pd.get_dummies(df['occupation'])

df_1 = pd.concat((df_1, dummy), axis=1)

df_1 = df_1.drop(['occupation'], axis=1)

df_1.head()

Unnamed: 0  age  fnlwgt  educational-num   relationship   race  \
0           0   25  226802                7      Own-child  Black   
1           1   38   89814                9        Husband  White   
2           2   28  336951               12        Husband  White   
3           3   44  160323               10        Husband  Black   
4           5   34  198693                6  Not-in-family  White   

   capital-gain  capital-loss  hours-per-week native-country  ...  \
0             0             0              40  United-States  ...   
1             0             0              50  United-States  ...   
2             0             0              40  United-States  ...   
3          7688             0              40  United-States  ...   
4             0             0              30  United-States  ...   

   Farming-fishing  Handlers-cleaners  Machine-op-inspct  Other-service  \
0                0                  0                  1              0   
1                1                  0                  0              0   
2                0                  0                  0              0   
3                0                  0                  1              0   
4                0                  0                  0              1   

   Priv-house-serv  Prof-specialty  Protective-serv  Sales  Tech-support  \
0                0               0                0      0             0   
1                0               0                0      0             0   
2                0               0                1      0             0   
3                0               0                0      0             0   
4                0               0                0      0             0   

   Transport-moving  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0  

[5 rows x 57 columns]

In [67]:
#OneHotEncoding relationship var
dummy = pd.get_dummies(df['relationship'])

df_1 = pd.concat((df_1, dummy), axis=1)

df_1 = df_1.drop(['relationship'], axis=1)

df_1.head()

Unnamed: 0  age  fnlwgt  educational-num   race  capital-gain  \
0           0   25  226802                7  Black             0   
1           1   38   89814                9  White             0   
2           2   28  336951               12  White             0   
3           3   44  160323               10  Black          7688   
4           5   34  198693                6  White             0   

   capital-loss  hours-per-week native-country  income  ...  Protective-serv  \
0             0              40  United-States       0  ...                0   
1             0              50  United-States       0  ...                0   
2             0              40  United-States       1  ...                1   
3             0              40  United-States       1  ...                0   
4             0              30  United-States       0  ...                0   

   Sales  Tech-support  Transport-moving  Husband  Not-in-family  \
0      0             0                 0        0              0   
1      0             0                 0        1              0   
2      0             0                 0        1              0   
3      0             0                 0        1              0   
4      0             0                 0        0              1   

   Other-relative  Own-child  Unmarried  Wife  
0               0          1          0     0  
1               0          0          0     0  
2               0          0          0     0  
3               0          0          0     0  
4               0          0          0     0  

[5 rows x 62 columns]

In [68]:
#OneHotEncoding race var
dummy = pd.get_dummies(df['race'])

df_1 = pd.concat((df_1, dummy), axis=1)

df_1 = df_1.drop(['race'], axis=1)

df_1.head()

Unnamed: 0  age  fnlwgt  educational-num  capital-gain  capital-loss  \
0           0   25  226802                7             0             0   
1           1   38   89814                9             0             0   
2           2   28  336951               12             0             0   
3           3   44  160323               10          7688             0   
4           5   34  198693                6             0             0   

   hours-per-week native-country  income  Federal-gov  ...  Not-in-family  \
0              40  United-States       0            0  ...              0   
1              50  United-States       0            0  ...              0   
2              40  United-States       1            0  ...              0   
3              40  United-States       1            0  ...              0   
4              30  United-States       0            0  ...              1   

   Other-relative  Own-child  Unmarried  Wife  Amer-Indian-Eskimo  \
0               0          1          0     0                   0   
1               0          0          0     0                   0   
2               0          0          0     0                   0   
3               0          0          0     0                   0   
4               0          0          0     0                   0   

   Asian-Pac-Islander  Black  Other  White  
0                   0      1      0      0  
1                   0      0      0      1  
2                   0      0      0      1  
3                   0      1      0      0  
4                   0      0      0      1  

[5 rows x 66 columns]

In [69]:
#OneHotEncoding native-country var
dummy = pd.get_dummies(df['native-country'])

df_1 = pd.concat((df_1, dummy), axis=1)

df_1 = df_1.drop(['native-country'], axis=1)

df_1.head()

Unnamed: 0  age  fnlwgt  educational-num  capital-gain  capital-loss  \
0           0   25  226802                7             0             0   
1           1   38   89814                9             0             0   
2           2   28  336951               12             0             0   
3           3   44  160323               10          7688             0   
4           5   34  198693                6             0             0   

   hours-per-week  income  Federal-gov  Local-gov  ...  Portugal  Puerto-Rico  \
0              40       0            0          0  ...         0            0   
1              50       0            0          0  ...         0            0   
2              40       1            0          1  ...         0            0   
3              40       1            0          0  ...         0            0   
4              30       0            0          0  ...         0            0   

   Scotland  South  Taiwan  Thailand  Trinadad&Tobago  United-States  Vietnam  \
0         0      0       0         0                0              1        0   
1         0      0       0         0                0              1        0   
2         0      0       0         0                0              1        0   
3         0      0       0         0                0              1        0   
4         0      0       0         0                0              1        0   

   Yugoslavia  
0           0  
1           0  
2           0  
3           0  
4           0  

[5 rows x 106 columns]

In [70]:
df_1 = df_1.drop(['Unnamed: 0'], axis = 1)

df_1.head()

age  fnlwgt  educational-num  capital-gain  capital-loss  hours-per-week  \
0   25  226802                7             0             0              40   
1   38   89814                9             0             0              50   
2   28  336951               12             0             0              40   
3   44  160323               10          7688             0              40   
4   34  198693                6             0             0              30   

   income  Federal-gov  Local-gov  Private  ...  Portugal  Puerto-Rico  \
0       0            0          0        1  ...         0            0   
1       0            0          0        1  ...         0            0   
2       1            0          1        0  ...         0            0   
3       1            0          0        1  ...         0            0   
4       0            0          0        1  ...         0            0   

   Scotland  South  Taiwan  Thailand  Trinadad&Tobago  United-States  Vietnam  \
0         0      0       0         0                0              1        0   
1         0      0       0         0                0              1        0   
2         0      0       0         0                0              1        0   
3         0      0       0         0                0              1        0   
4         0      0       0         0                0              1        0   

   Yugoslavia  
0           0  
1           0  
2           0  
3           0  
4           0  

[5 rows x 105 columns]

As shown in the output above, all categorical variables and respective levels were transformed into binary form.

In [71]:
### verify all variables have correct type

df_1.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45175 entries, 0 to 45174
Data columns (total 105 columns):
 #   Column                      Dtype
---  ------                      -----
 0   age                         int64
 1   fnlwgt                      int64
 2   educational-num             int64
 3   capital-gain                int64
 4   capital-loss                int64
 5   hours-per-week              int64
 6   income                      uint8
 7   Federal-gov                 uint8
 8   Local-gov                   uint8
 9   Private                     uint8
 10  Self-emp-inc                uint8
 11  Self-emp-not-inc            uint8
 12  State-gov                   uint8
 13  Without-pay                 uint8
 14  Female                      uint8
 15  Male                        uint8
 16  10th                        uint8
 17  11th                        uint8
 18  12th                        uint8
 19  1st-4th                     uint8
 20  5th-6th                    

It's important to ensure all variables have the right type for memory and modeling purposes.

### Split Dataset

First, we seperate the entire dataset into 'train' and 'test' with a 90/10 split. This 'test' set will be used as a hold out set for final evaluation after models have been hyperparameter tuned.

In [72]:
# create train, validation, and test sets

from sklearn.model_selection import train_test_split
from collections import Counter

df_2 = df_1.copy()

train, test = np.split(df_2.sample(frac=1), [int(.9*len(df_2))])
print(len(train))


40657


In [73]:
### import relevant function
from sklearn.model_selection import ShuffleSplit

### acquire target label values, delete column of said acquired values, and assign remaining variables and corresponding values to 'X'

if 'income' in train:
    y_train = train['income'].values 
    del train['income'] 
    X_train = train.values 


# target label values for test set
if 'income' in test:
    y_test = test['income'].values 
    del test['income'] 
    X_test = test.values 


### Sampling

As established above, given the unbalanced distribution of the target variable, we will be using different sampling methods to increase performance of models.

In [74]:
# Over sample dataset
from imblearn.over_sampling import RandomOverSampler
over_sampler = RandomOverSampler(random_state=42)
X_over, y_over = over_sampler.fit_resample(X_train, y_train)
print(f"Training target statistics: {Counter(y_over)}")

Training target statistics: Counter({1: 30523, 0: 30523})


In [75]:
# Under sample
from imblearn.under_sampling import RandomUnderSampler
under_sampler = RandomUnderSampler(random_state=42)
X_under, y_under = under_sampler.fit_resample(X_train, y_train)
print(f"Training target statistics: {Counter(y_under)}")

Training target statistics: Counter({0: 10134, 1: 10134})


### Standardize and Scale Data

We will standardize the dataset to ensure that each variable is weighted equally, since there are many variables on very different scales as well as one hot encoded categorical variables. This should help reduce model bias, improving performance, as well as making feature importance easier to interpret later on. 

In [76]:
#standard set
scl_obj = StandardScaler()
scl_obj.fit(X_train)
X_train = scl_obj.transform(X_train) 
X_test = scl_obj.transform(X_test)

#oversampled data 
scl_obj = StandardScaler()
scl_obj.fit(X_over)
X_over = scl_obj.transform(X_over) 
X_test_over = scl_obj.transform(X_test)

#undersampled data
scl_obj = StandardScaler()
scl_obj.fit(X_under)
X_under = scl_obj.transform(X_under) 
X_test_under = scl_obj.transform(X_test)



##  Data Preparation, Part 2

Describe the final dataset that is used for classification/regression (include a description of any newly formed variables you created).

Here is a glimpse of how the dataset looks like:

In [77]:
print(train.head())
print(train.shape)

       age  fnlwgt  educational-num  capital-gain  capital-loss  \
12859   38  107410                5             0             0   
25762   47   33710               13          3103             0   
3625    46   35136                7             0             0   
16840   39  165186               10             0             0   
40673   38  105150               13             0             0   

       hours-per-week  Federal-gov  Local-gov  Private  Self-emp-inc  ...  \
12859              45            0          0        0             0  ...   
25762              60            0          0        1             0  ...   
3625               40            1          0        0             0  ...   
16840              45            0          0        1             0  ...   
40673              45            0          0        1             0  ...   

       Portugal  Puerto-Rico  Scotland  South  Taiwan  Thailand  \
12859         0            0         0      0       0         0   


In the 'train' dataset, there are 40657 instances, wherein each instance represents an individual. Due to the transformation of categorical variables to OneHotEncoding binary format, the dataset holds 104 attributes for each instance. The value of the target variable is represented as a 1 if the individual has a reported annual income of greater than or equal to 50K USD and a 0 if it is less than 50K USD.

Additionally, given the unbalanced distribution of the target variable, we appplied two sampling methods (over and under) that may increase performance and evaluation of models being trained on this data.

# Modeling and Evaluation

## Evaluation Metrics

Choose and explain your evaluation metrics that you will use (i.e., accuracy,
precision, recall, F-measure, or any metric we have discussed). Why are the measure(s) appropriate for analyzing the results of your modeling? Give a detailed explanation backing up any assertions.

### Income Classification Evaluation Metrics
Accuracy: <br>
> The first metric we will be using is Accuracy. By accuracy, we mean "the ratio of number of correct predictions to the total number of input samples."[^1] 

> Accuracy = Number of Correct Predictions/Total Number of Predictions Made

> By balancing our datset with respect to our target variable, we took measures to maximize the value that this metric provides. For our dataset, accuracy will give us a measure of how maany instances are correctly classified as >=50K  or <50K, in any given model.

Precision: <br>
> In order to get a more detailed look at what accuracy represents, we first break down the metric into Precision and Recall. Precision, in essence, outlines the proportion of positive identifications was actually correct[^2].

> Precision = True Positive/(True Positive + False Positive)

> For our dataset, Precision will give us a measure of the proportion of positives (>=50K) were classified correctly.

Recall: <br>

> Recall, on the other hand, measures the proportion of actual positives that  were identified correctly[^3].

> Recall = True Positive/(True Positive + False Negative)

> For our dataset, Recall will give us a measure of how many of the instances that were actually >=50K were classified as such by the model.

Sensitivity: <br>
> Sensitivity (True Positive Rate) refers to the "proportion of positive data points that are correctly considered as positive, with respect to all positive data points."[^4]

> Sensitivity = True Positive/(False Negative + True Positive)

> For our dataset, sensitivity measures the proportion of actual >=50K classified as such by a given model.

Specificity: <br>
> Inversely to the previous measure, specificity (True Negative Rate) refers to the "proportion of negative data points that are correctly considered as negative, with respect to all negative data points."[^5]

> Sensitivity = True Negative/(True Negative + False Positive)

> For our dataset, specificity measures the proportion of actual <50K classified as such by a given model.

While we will look at the above metrics for income classifcation, the primary metric we will be looking to optimize in this case is precision, as it is the more useful metric in many contexts because it is harder to predict because it is a minority of people, and because predicting high earners could be used for things like targeted marketing campaigns for high margin products. 

[^1]: https://towardsdatascience.com/metrics-to-evaluate-your-machine-learning-algorithm-f10ba6e38234
[^2]: https://developers.google.com/machine-learning/crash-course/classification/precision-and-recall
[^3]: https://developers.google.com/machine-learning/crash-course/classification/precision-and-recall
[^4]: https://towardsdatascience.com/metrics-to-evaluate-your-machine-learning-algorithm-f10ba6e38234
[^5]: https://towardsdatascience.com/metrics-to-evaluate-your-machine-learning-algorithm-f10ba6e38234

### Age Prediction Evaluation Metric

## Splitting

Choose the method you will use for dividing your data into training and
testing splits (i.e., are you using Stratified 10-fold cross validation? Why?). Explain why
your chosen method is appropriate or use more than one method as appropriate. For example, if you are using time series data then you should be using continuous training and testing sets across time.

The primary method used for splitting between training and testing splits is 10 fold shuffle cross validation. Shuffle split was chosen over K fold CV to help keep the validation scalable if the datasize is to grow in the future (for example if more years are to be brought in). It is not a large concern if a single observation is used multiple times, or one is not used at all, because with the current size of this dataset, the risk of over-biasing the model because of that is very low. 

With inbalanced data such as this, biasing for the majority class is a large risk. Random over and under sampling techinques were decided to be used to attempt to deal with bias from the lack of balance in the dataset rather than a stratified validation method. The over sampling method in particular is likely to help improve our important precision metric by intentionally introducing bias towards the minority class (income > 50k) by randomly duplicating minority class observations to even out the balance of the dataset. The undersampling method instead seeks to reduce bias by randomly removing the majority class, however this comes with a higher risk of removing important information by cutting down the size of the data by a large margin, and because of this is the less preferred method. 

### Split Dataset

In data preparation, we seperated the entire dataset into 'train' and 'test' with a 90/10 split. This 'test' set will be used for final evaluation after models have been hyperparameter tuned.

Within the 'train' set, we create a 10-fold shuffle split cross validation object that can be used with the regular, over, and under sampled training data.

In [78]:
### import relevant function
from sklearn.model_selection import ShuffleSplit

### split data 80/20 with 10 cross-validation
num_cv_iterations = 10
cv_object = ShuffleSplit(n_splits=num_cv_iterations,
                         test_size  = 0.2, random_state = 42)
                         
print(cv_object)

ShuffleSplit(n_splits=10, random_state=42, test_size=0.2, train_size=None)


## Model Training

Create three different classification/regression models for each task (e.g., random forest, KNN, and SVM for task one and the same or different algorithms for task two). Two modeling techniques must be new (but the third could be SVM or logistic regression). Adjust parameters as appropriate to increase generalization performance using your chosen metric. You must investigate different parameters of the algorithms!

### Income Classification

In [79]:
%%time
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import BernoulliNB
from sklearn.model_selection import cross_validate

### create LR object and apply parameters and consants

num_estimators = 50
# lets train some trees
clf_array = [
    ('Random Trees 1 per leaf', RandomForestClassifier(random_state = 42, min_samples_leaf = 1, n_estimators = num_estimators)),
    ('Random Trees 10 per leaf', RandomForestClassifier(random_state = 42, min_samples_leaf = 10, n_estimators = num_estimators)),
    ('Random Trees 30 per leaf', RandomForestClassifier(random_state = 42, min_samples_leaf = 30, n_estimators = num_estimators)),
    ('Random Trees 100 per leaf', RandomForestClassifier(random_state = 42, min_samples_leaf = 100, n_estimators = num_estimators)),
    ('LogReg L2 C1', LogisticRegression(penalty='l2', C=1.0, class_weight=None, solver='liblinear' )),
    ('LogReg L1 C1', LogisticRegression(penalty='l1', C=1.0, class_weight=None, solver='liblinear' )),
    ('LogReg L2 C.1', LogisticRegression(penalty='l2', C=.1, class_weight=None, solver='liblinear' )),
    ('LogReg L1 C.1', LogisticRegression(penalty='l1', C=.1, class_weight=None, solver='liblinear' )),
    ('LogReg L2 C5', LogisticRegression(penalty='l2', C=5.0, class_weight=None, solver='liblinear' )),
    ('LogReg L1 C5', LogisticRegression(penalty='l1', C=5.0, class_weight=None, solver='liblinear' )),
    ('Gaussian Naive Bayes', GaussianNB()),
    ('Bernoulli Naive Bayes', BernoulliNB())
    ]

scoring = {'acc': 'accuracy',
           'prec': 'precision'}

for clf in clf_array:
    scores = cross_validate(clf[1], X_train, y_train, scoring=scoring,
                         cv=cv_object, return_train_score=False)
    scoreso = cross_validate(clf[1], X_over, y_over, scoring=scoring,
                         cv=cv_object, return_train_score=False)
    scoresu = cross_validate(clf[1], X_under, y_under, scoring=scoring,
                         cv=cv_object, return_train_score=False)
    print ("Model:", clf[0], "Accuracy (Regular, Oversample, Undersample):", round(scores['test_acc'].mean(),2), round(scoreso['test_acc'].mean(),2), round(scoresu['test_acc'].mean(),2),
    "Precision (Regular, Oversample, Undersample):", round(scores['test_prec'].mean(),2), round(scoreso['test_prec'].mean(),2), round(scoresu['test_prec'].mean(),2)) 

Model: Random Trees 1 per leaf Accuracy (Regular, Oversample, Undersample): 0.85 0.93 0.81 Precision (Regular, Oversample, Undersample): 0.73 0.9 0.81
Model: Random Trees 10 per leaf Accuracy (Regular, Oversample, Undersample): 0.86 0.84 0.82 Precision (Regular, Oversample, Undersample): 0.78 0.81 0.79
Model: Random Trees 30 per leaf Accuracy (Regular, Oversample, Undersample): 0.85 0.83 0.82 Precision (Regular, Oversample, Undersample): 0.78 0.8 0.79
Model: Random Trees 100 per leaf Accuracy (Regular, Oversample, Undersample): 0.84 0.82 0.8 Precision (Regular, Oversample, Undersample): 0.77 0.8 0.77
Model: LogReg L2 C1 Accuracy (Regular, Oversample, Undersample): 0.85 0.82 0.82 Precision (Regular, Oversample, Undersample): 0.73 0.81 0.8
Model: LogReg L1 C1 Accuracy (Regular, Oversample, Undersample): 0.85 0.82 0.82 Precision (Regular, Oversample, Undersample): 0.73 0.81 0.8
Model: LogReg L2 C.1 Accuracy (Regular, Oversample, Undersample): 0.85 0.82 0.82 Precision (Regular, Oversample,

c:\Users\dobbs\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
Traceback (most recent call last):
  File "c:\Users\dobbs\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 531, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "c:\Users\dobbs\anaconda3\lib\site-packages\sklearn\naive_bayes.py", line 641, in fit
    self._count(X, Y)
  File "c:\Users\dobbs\anaconda3\lib\site-packages\sklearn\naive_bayes.py", line 763, in _count
    check_non_negative(X, "MultinomialNB (input X)")
  File "c:\Users\dobbs\anaconda3\lib\site-packages\sklearn\utils\validation.py", line 1046, in check_non_negative
    raise ValueError("Negative values in data passed to %s" % whom)
ValueError: Negative values in data passed to MultinomialNB (input X)

  warnings.warn("Estimator fit failed. The score on this trai

Model: Multinomial Naive Bayes Accuracy (Regular, Oversample, Undersample): nan nan nan Precision (Regular, Oversample, Undersample): nan nan nan
Model: Bernoulli Naive Bayes Accuracy (Regular, Oversample, Undersample): 0.78 0.79 0.78 Precision (Regular, Oversample, Undersample): 0.54 0.76 0.75
Wall time: 17min 22s


### Age Prediction

## Results Analysis and Visualization



Analyze the results using your chosen method of evaluation. Use visualizations of the results to bolster the analysis. Explain any visuals and analyze why they are interesting to someone that might use this model.

### Income Classification

In [ ]:
weights = []
# run the pipline cross validated
for iter_num, (train_indices, test_indices) in enumerate(cv_object.split(X,y)):
    piped_object.fit(X[train_indices],y[train_indices])  # train object
    # it is a little odd getting trained objects from a  pipeline:
    weights.append(piped_object.named_steps['logit_model'].coef_[0])
    

weights = np.array(weights)

In [ ]:
import plotly
plotly.offline.init_notebook_mode() # run at the start of every notebook

error_y=dict(
            type='data',
            array=np.std(weights,axis=0),
            visible=True
        )

graph1 = {'x': df_imputed.columns,
          'y': np.mean(weights,axis=0),
    'error_y':error_y,
       'type': 'bar'}

fig = dict()
fig['data'] = [graph1]
fig['layout'] = {'title': 'Logistic Regression Weights, with error bars'}

plotly.offline.iplot(fig)

### Age Prediction

## Model Advtanages

Discuss the advantages of each model for each classification task, if any. If there are not advantages, explain why. Is any model better than another? Is the difference significant with 95% confidence? Use proper statistical comparison methods. You must use statistical comparison techniques—be sure they are appropriate for your chosen method of validation as discussed in unit 7 of the course.

### Income Classification

### Age Prediction

## Attribute Importance

Which attributes from your analysis are most important? Use proper methods discussed in class to evaluate the importance of different attributes. Discuss the results and hypothesize about why certain attributes are more important than others for a given classification task.

### Income Classification

### Age Prediction

## Deployment

How useful is your model for interested parties (i.e., the companies or organizations that might want to use it for prediction)? How would you measure the model's value if it was used by these parties? How would your deploy your model for interested parties? What other data should be collected? How often would the model need to be updated, etc.? 

## Exceptional Work

You have free reign to provide additional analyses. One idea: grid search parameters in a parallelized fashion and visualize the performances across attributes. Which parameters are most significant for making a good model for each classification algorithm?